server/src/endpoint/writes.js

```diff
@@ -6,46 +6,135 @@ const r = require('rethinkdb');
-      check(results.changes.length > results_index);
-      const res = results.changes[results_index++];
-      if (res.error !== undefined) {
-        if (res.error.indexOf('Duplicate primary key') === 0) {
-          data[i] = { error: 'The document already exists.' };
+const invalidated_msg = 'Write invalidated by another request, try again.';
+const missing_msg = 'The document was missing.';
+const timeout_msg = 'Operation timed out.';
+const unauthorized_msg = 'Operation not permitted.';
+
+const hz_v = '$hz_v$';
+const apply_version = (row, new_version) => row.merge(r.object(hz_v, new_version));
+
+const make_write_response = (data) => {
+  data.forEach((item, index) => {
+    if (item instanceof Error) {
+      data[index] = { error: item.message };
+    }
+  });
+  return { data, state: 'complete' };
+};
+
+// This function returns a Promise that resolves to an array of responses - one for each row in
+//  `original_rows`, or rejects with an appropriate error.
+// timeout -> integer
+//   minimum number of milliseconds before giving up on retrying writes
+//   null means no timeout
+// pre_validate -> function (rows):
+//   rows: all pending rows
+//   return: a (promise of an) array of info for those rows (which will be passed to the validate step)
+// validate_row -> function (row, info):
+//   row: The row from the original query
+//   info: The info returned by the pre_validate step for this row
+//   return: nothing if successful or an error to be put as the response for this row
+// do_write -> function (rows):
+//   rows: all pending rows
+//   return: a (promise of a) ReQL write result object
+const retry_loop = (original_rows, ruleset, timeout, pre_validate, validate_row, do_write) => {
+  const response_data = Array(original_rows.length).fill(null);
+
+  // Save original version fields, which may get clobbered, so we don't have to copy everything
+  let row_data = original_rows.map((row, index) => ({ row, index, version: row[hz_v] }));
+  let deadline;
+
+  const iterate = () => {
+    if (row_data.length === 0) {
+      return response_data;
+    } else if (timeout !== null) {
+      if (!deadline) {
+        deadline = Date.now() + timeout;
+      } else if (Date.now() > deadline) {
+        response_data.forEach((data, index) => {
+          if (data === null) {
+            response_data[index] = new Error(timeout_msg);
+          }
+        });
+        return response_data;
+      }
+    }
+
+    return Promise.resolve().then(() => {
+      // Gather all rows to write
+      row_data.forEach((data) => {
+        if (data.version === undefined) {
+          delete data.row[hz_v];
         } else {
-          data[i] = { error: res.error };
-        }
-      } else if (res.new_val === null) {
-        data[i] = { id: res.old_val.id };
-        if (res.old_val[version_field] !== undefined) {
-          data[i][version_field] = res.old_val[version_field];
+          data.row[hz_v] = data.version;
```

Comment:
I'm still a little confused.  Why would the validator set the version on the row?  What does it mean if it sets it to a particular value?  And why is it OK to discard that value here?

(It might be nice to have that in a comment, actually, since it's kind of non-obvious why the validators would change this and what they change it to semantically.)